hail/src/main/resources/include/hail/NDArray.h

```diff
@@ -0,0 +1,72 @@
+#ifndef HAIL_NDARRAY_H
+#define HAIL_NDARRAY_H 1
+
+#include <vector>
+
+struct NDArray {
+  int flags; // least sig. bit denotes if row major
+  size_t elem_size;
+  std::vector<long> shape;
+  std::vector<long> strides;
+  const char *data;
+};
+
+NDArray make_ndarray(int flags, size_t elem_size, std::vector<long> shape, const char *data);
+char const *load_ndarray_addr(NDArray nd, std::vector<long> indices);
+
+void set_strides_row_major(std::vector<long> &strides, std::vector<long> &shape);
+void set_strides_col_major(std::vector<long> &strides, std::vector<long> &shape);
+
+NDArray make_ndarray(int flags, size_t elem_size, std::vector<long> shape, const char *data) {
```

Comment:
last comment (sorry!)

I don't see a check that enforces that the data array is the right size for the shape. Can you add one? I'm not sure if you want to enforce that it's exactly the right size or if it can be bigger and only the relevant bit is used. It should definitely error if it's too small, I think.

Reply:
Good idea, will do!